### PR TITLE
Data loader

### PR DESCRIPTION
**Pull request by MekWarrior**
This PR adds an elementary data loader that is built to read from the CellNet data ontology. It allows for human-readable calls to build raw data and packages it with the accompanying metadata files to prepare it for submission to crowdsourcing platforms. 

It represents the first step toward a complete overhaul of our data pipeline, and, as such, will undergo several changes once DeepCell Datasets is closer to being a mature repo.

There are also miscellaneous critical updates to inits and testing that bring the current toolbox repo closer to maturity.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.6
+FROM python:3.7
```
Comment (willgraf) [inline Dockerfile:1]: Not an issue, just wondering why this is necessary

File: caliban_toolbox/__init__.py [hunk 2]
```diff
@@ -23,4 +23,25 @@
...
+from caliban_toolbox import utils
+from caliban_toolbox import aws_functions
+from caliban_toolbox import figure_eight_functions
+from caliban_toolbox import log_file
+from caliban_toolbox import relabel
+from caliban_toolbox import reshape_data
+
+from caliban_toolbox.pre_annotation import *
+# from caliban_toolbox.tracking import *
+from caliban_toolbox.utils import *
```
Comment (willgraf) [inline caliban_toolbox/__init__.py:43]: Prefer to upload by name, or else it's possible to import 3rd party packages from that module.

You can also use `__all__` in the file to explicitly define what is able to be imported from a file.

File: requirements.txt [hunk 3]
```diff
@@ -1,7 +1,8 @@
 pandas>=0.23.3,<1
-numpy>=1.14.5,<2
+pymongo>=3.10.1
```
Comment (willgraf) [inline requirements.txt:2]: is pymongo used in this PR?

File: caliban_toolbox/pre_annotation/data_loader.py [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+        self._vocab_check()
+
+        self.base_path = '/data/raw_data'
+        for item in self.data_type:
+            self.base_path = os.path.join(self.base_path, item)
+
+        self._datasets_available()  # TODO: keep list of datasets for comparison
+        self._calc_upper_bound()
+
+        # self.mng_db = self._setup_mongo()
```
Comment (willgraf) [inline caliban_toolbox/pre_annotation/data_loader.py:110]: I think we should not include the pymongo functions at all yet. let's just prune them out of this PR and add them in the next version.

File: caliban_toolbox/pre_annotation/data_loader.py [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+        new_compartments = []
+        for item in self.compartments:
+            if any([item.lower() == 'nuc',
+                    item.lower() == 'nuclear']):
+                new_compartments.append('Nuclear')
+            elif any([item.lower() == 'wholecell',
+                      item.lower() == 'whole_cell']):
+                new_compartments.append('WholeCell')
+            else:
+                new_compartments.append(item)
```
Comment (willgraf) [inline caliban_toolbox/pre_annotation/data_loader.py:142]: Could clean this up by first setting `item = item.lower()`, then just checking if `item in {'nuc', 'nuclear'}`. A bit easier to extend in the future that way I think.

File: caliban_toolbox/pre_annotation/data_loader_test.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        # test with standard inputs
+        _ = data_loader.UniversalDataLoader(data_type=loader_inputs[0],
+                                            imaging_types=loader_inputs[1],
+                                            specimen_types=loader_inputs[2],
+                                            compartments=loader_inputs[3],
+                                            markers=loader_inputs[4],
+                                            exp_ids=loader_inputs[5],
+                                            sessions=loader_inputs[6],
+                                            positions=loader_inputs[7],
+                                            file_type=loader_inputs[8])
```
Comment (willgraf) [inline caliban_toolbox/pre_annotation/data_loader_test.py:80]: Can we do anything with this instance? The tests are a little flimsy right now.

File: Dockerfile (hunk 1 above)
Comment (MekWarrior) [reply Dockerfile:1]: No issue. Upgraded to match Caliban.

File: requirements.txt (hunk 3 above)
Comment (MekWarrior) [reply requirements.txt:2]: Not in this PR, but will be in the next once DeepCell Datasets is up and running

File: caliban_toolbox/pre_annotation/data_loader_test.py (hunk 6 above)
Comment (MekWarrior) [reply caliban_toolbox/pre_annotation/data_loader_test.py:80]: Yes, the tests need to be bulked up. We also don't know is going to stay from this "patch." Testing is also difficult because it would need to run on a mounted volume.

File: caliban_toolbox/__init__.py (hunk 2 above)
Comment (ngreenwald) [reply caliban_toolbox/__init__.py:43]: As you no doubt noticed, I didn't include any imports for anything. I'm not sure I quite understand the point of doing this? Is it good software practice to explicitly define what can be imported to avoid having things you don't want accessible? Or is there some other reason that it's good to include these?

Comment (willgraf) [reply caliban_toolbox/__init__.py:43]: Yeah its just improving/cleaning the namespace of the API. Its weird to have `caliban.pre_annotation.skimage` in the namespace.

Comment (MekWarrior) [reply caliban_toolbox/__init__.py:43]: will adjust in future PR when the utils and data loader functions are more set

File: caliban_toolbox/pre_annotation/data_loader.py (hunk 5 above)
Comment (MekWarrior) [reply caliban_toolbox/pre_annotation/data_loader.py]: done and added dicts to keep track of future misspellings

File: caliban_toolbox/pre_annotation/data_loader.py [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+        sessions (list): which sessionss to include
+        positions/FOVs (list): which positionss/FOVs to include
+
+            - list should be strings and match CellNet ontology
+              (e.g. data type = ['dynamic', '2d'])
+            - 'all' selects all data from a given catagory
+              (e.g. sessions=['all'])
+
+    Returns:
+        Numpy array with the shape [fovs, tifs, y_dim, x_dim]
```
Comment (ngreenwald) [inline caliban_toolbox/pre_annotation/data_loader.py:72]: Is tifs the same as markers/channels?

File: caliban_toolbox/pre_annotation/data_loader.py [hunk 8]
```diff
@@ -0,0 +1,465 @@
...
+
+    def _vocab_check(self):
+        """Check each user input for common mistakes and correct as neccesary
+        """
+        # TODO: improve this for generality and scale
+
+        # Dictionaries of common spellings
+        imaging_fluo_dict = {'flourescent', 'fluorescence', 'fluorescent', 'fluo'}
+        comp_nuc_dict = {'nuc', 'nuclear'}
+        comp_wc_dict = {'wholecell', 'whole_cell', }
```
Comment (willgraf) [inline caliban_toolbox/pre_annotation/data_loader.py:120]: technically they are sets, not dicts. maybe `bucket` or `names` or something?

File: caliban_toolbox/pre_annotation/data_loader.py (hunk 7 above)
Comment (MekWarrior) [reply caliban_toolbox/pre_annotation/data_loader.py:72]: I don't know what you mean by markers/channels, you'll need to be more specific. Our ontology stores images according to static/dynamic, 2d/3d, imaging type, specimen type, compartment_marker (eg. nuclear_dapi), and experiment ID. The files in this final directory are expected to be named according to session and FOV/position. So there are no "marker/channels" to return. The metadata files keeps a record of whats in each row of the array for reference.

Comment (ngreenwald) [reply caliban_toolbox/pre_annotation/data_loader.py:72]: Ah, sorry. My question is what doe the TIFs axis represent here. Is this whatever extra dimension the data has? For example, for 3D data, is TIFs the z-dimension, and there isn't a channels dimension, whereas for multichannel data TIFs is the channels dimension?

File: caliban_toolbox/pre_annotation/data_loader.py (hunk 8 above)
Comment (MekWarrior) [reply caliban_toolbox/pre_annotation/data_loader.py:120]: It's dictionary of misspellings for a given word. I've changed name to avoid confusion in the interim.

File: caliban_toolbox/pre_annotation/data_loader.py (hunk 4 above)
Comment (MekWarrior) [reply caliban_toolbox/pre_annotation/data_loader.py]: Done

File: caliban_toolbox/pre_annotation/data_loader.py (hunk 7 above)
Comment (MekWarrior) [reply caliban_toolbox/pre_annotation/data_loader.py]: There is no multichannel data for the function to load, as that is handled by virtue of the ontology - each channel has its own sub directory.  

I have amended the docstring to make it clear that the tifs dim is a z-dim that either represents space or time. This is temporary as we will need to accommodate 3D dynamic data as well, but that is easily solved with another sql lookup and dimension.
